misk-hibernate/src/main/kotlin/misk/hibernate/actions/HibernateDatabaseQueryDynamicAction.kt

```diff
@@ -0,0 +1,122 @@
+package misk.hibernate.actions
+
+import com.google.inject.Injector
+import misk.MiskCaller
+import misk.exceptions.BadRequestException
+import misk.exceptions.UnauthorizedException
+import misk.hibernate.DbEntity
+import misk.hibernate.Query
+import misk.hibernate.ReflectionQuery
+import misk.hibernate.Session
+import misk.hibernate.Transacter
+import misk.hibernate.actions.HibernateDatabaseQueryWebActionModule.Companion.checkDynamicQuery
+import misk.hibernate.actions.HibernateDatabaseQueryWebActionModule.Companion.findDatabaseQueryMetadata
+import misk.hibernate.actions.HibernateDatabaseQueryWebActionModule.Companion.getTransacterForDatabaseQueryAction
+import misk.scope.ActionScoped
+import misk.web.Post
+import misk.web.RequestBody
+import misk.web.RequestContentType
+import misk.web.ResponseContentType
+import misk.web.actions.WebAction
+import misk.web.dashboard.AdminDashboardAccess
+import misk.web.mediatype.MediaTypes
+import misk.web.metadata.DatabaseQueryMetadata
+import javax.inject.Inject
+import javax.inject.Singleton
+import kotlin.reflect.KClass
+import kotlin.reflect.full.memberProperties
+
+/** Runs query from Database Query dashboard tab against DB and returns results */
+@Singleton
+class HibernateDatabaseQueryDynamicAction @Inject constructor(
+  @JvmSuppressWildcards private val callerProvider: ActionScoped<MiskCaller?>,
+  val databaseQueryMetadata: List<DatabaseQueryMetadata>,
+  val queries: List<HibernateQuery>,
+  val injector: Injector,
+) : WebAction {
+
+  val maxMaxRows = 500
+  val rowCountErrorLimit = 30
+  val rowCountWarningLimit = 20
+  private val queryFactory = ReflectionQuery.Factory(ReflectionQuery.QueryLimitsConfig(
+      maxMaxRows, rowCountErrorLimit, rowCountWarningLimit))
+
+  @Post(HIBERNATE_QUERY_DYNAMIC_WEBACTION_PATH)
+  @RequestContentType(MediaTypes.APPLICATION_JSON)
+  @ResponseContentType(MediaTypes.APPLICATION_JSON)
+  @AdminDashboardAccess
+  fun getAll(@RequestBody request: Request): Response {
+    val caller = callerProvider.get()!!
+    val queryClass = request.queryClass
+
+    checkDynamicQuery(queryClass, true)
+
+    val metadata = findDatabaseQueryMetadata(databaseQueryMetadata, queryClass)
+    val transacter = getTransacterForDatabaseQueryAction(injector, metadata)
+
+    val results = if (caller.isAllowed(metadata.allowedCapabilities, metadata.allowedServices)) {
```

Comment:
My first instinct is that this should be admins only? and not more granular than that?

Reply:
Providing granularity would allow the case where a service owner wants to open up a read only query to the broader organization for debugging purposes but lock down all other queries to admins.